File: generated-code/haskell/codecraft/TcpReadWrite/src/Model/DebugInterface/PrimitiveType.hs
module Model.DebugInterface.PrimitiveType where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int

-- | Primitive type for debug rendering
data PrimitiveType
    -- | Lines, number of vertices should be divisible by 2
    = Lines
    -- | Triangles, number of vertices should be divisible by 3
    | Triangles
    deriving (Eq, Ord, Show)

instance Trans PrimitiveType where
    read = do
        tag :: Int32 <- Trans.read
        return $ case tag of
            0 -> Lines
            1 -> Triangles

    write Lines =
        Trans.write (0 :: Int32)
    write Triangles =
        Trans.write (1 :: Int32)

File: generated-code/php/codecraft/TcpReadWrite/Model/DebugInterface/PrimitiveType.php
<?php

namespace Model\DebugInterface {
    require_once 'Stream.php';

    /**
     * Primitive type for debug rendering
     */
    abstract class PrimitiveType
    {
        /**
         * Lines, number of vertices should be divisible by 2
         */
        const LINES = 0;

        /**
         * Triangles, number of vertices should be divisible by 3
         */
        const TRIANGLES = 1;

        /**
         * Read PrimitiveType from input stream
         */
        public static function readFrom(\InputStream $stream): int
        {
            $result = $stream->readInt32();
            if (0 <= $result && $result < 2) {
                return $result;
            }
            throw new Exception('Unexpected tag value');
        }
    }
}